### PR TITLE
[DNM] neorados: add more unit tests

### PR DESCRIPTION
**Pull request by yuvalif**
please review only: test_neorados.cc
the rest of the changes are taken from: https://github.com/ceph/ceph/pull/48129

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/test/neorados/test_neorados.cc [hunk 1]
```diff
@@ -29,11 +54,484 @@ TEST_F(TestNeoRADOS, MakeWithLibRADOS) {
...
+
+  // provide pool that doesn't exists -- just testing round-trip
+  ASSERT_THROW(
+      rados.execute({"dummy-obj"}, IOContext(std::numeric_limits<int64_t>::max()),
+                  std::move(op), &bl, ceph::async::use_blocked),
+      boost::system::system_error);
+}
+
+TEST_F(TestNeoRADOS, MakeWithBuilder) {
+  GTEST_SKIP(); // the build command hangs with the "use_blocked" flag
```
Comment (yuvalif) [inline src/test/neorados/test_neorados.cc:79]: currently, it is not possible to use the "Builder" together with "use_blocked"

File: src/test/neorados/test_neorados.cc [hunk 2]
```diff
@@ -29,11 +54,484 @@ TEST_F(TestNeoRADOS, MakeWithLibRADOS) {
...
+
+  // provide pool that doesn't exists -- just testing round-trip
+  ASSERT_THROW(
+      rados.execute({"dummy-obj"}, IOContext(std::numeric_limits<int64_t>::max()),
+                  std::move(op), &bl, boost::asio::use_future).get(),
+      boost::system::system_error);
+}
+
+TEST_F(TestNeoRADOSAsync, MakeWithBuilder) {
+  GTEST_SKIP(); // the build command hangs with the "use_future" flag
```
Comment (yuvalif) [inline src/test/neorados/test_neorados.cc:223]: currently, it is not possible to use the "Builder" together with "use_future"

File: src/test/neorados/test_neorados.cc [hunk 3]
```diff
@@ -29,11 +54,484 @@ TEST_F(TestNeoRADOS, MakeWithLibRADOS) {
...
+
+  const std::set<std::string> pools{"piscine", "piscina", "zwembad"};
+  std::set<std::string> fetched_pools;
+
+  boost::asio::co_spawn(ctx, 
+    [&]() -> boost::asio::awaitable<void> {
+      auto rados = co_await RADOS::Builder{}.build(ctx, boost::asio::use_awaitable);
+      auto pool_deleter = make_scope_guard([&]() {
+          for (const auto& pool: pools) {
+            rados.delete_pool(pool, boost::asio::detached);
```
Comment (yuvalif) [inline src/test/neorados/test_neorados.cc:506]: this seems to delete only one of the pools

File: src/test/neorados/test_neorados.cc [hunk 4]
```diff
@@ -29,11 +54,484 @@ TEST_F(TestNeoRADOS, MakeWithLibRADOS) {
...
+  std::set<std::string> objects;
+  std::set<std::string> fetched_objects;
+  fill_objects(1, 100, objects);
+
+  boost::asio::co_spawn(ctx, 
+    [&]() -> boost::asio::awaitable<void> {
+      auto rados = co_await RADOS::Builder{}.build(ctx, boost::asio::use_awaitable);
+      const std::string pool_id{"piscine"};
+      auto pool_deleter = make_scope_guard([&]() {
+          rados.delete_pool(pool_id, boost::asio::detached);
```
Comment (yuvalif) [inline src/test/neorados/test_neorados.cc:450]: this does not always delete the pool

File: src/test/neorados/test_neorados.cc (hunk 3 above)
Comment (adamemerson) [reply src/test/neorados/test_neorados.cc:506]: Yes. I think fundamentally the problem is that once the destructor exits things fundamentally get deleted and without the destructor being a coroutine there's no good way to wait, especially in a single threaded model.

I've got https://github.com/ceph/ceph/pull/48234/commits/3c3732625a6afdaefdd213b9dd2e27621ce53843#diff-e7a05e42b0b2a31b55db86bb74d1c3cfe3c664730c8b939383a90ca945f0ada9R65 in progress that I think might do a better job for coroutine tests.

Comment (yuvalif) [reply src/test/neorados/test_neorados.cc:506]: my concern is not about the unit tests here (for that I can just delete all pools in blocking calls after the test).
the main goal is to understand how to write exception-safe code with coroutines.

Comment (cbodley) [reply src/test/neorados/test_neorados.cc:506]: by the time we get to an object's destructor, it's too late for any async cleanup. RAII works fine for stuff like memory allocations, but any async cleanup needs to be part of the async control flow

for graceful shutdowns, maybe take a look at the asio frontend for an example? in https://github.com/ceph/ceph/blob/main/src/rgw/rgw_asio_frontend.cc#L1063-L1091 we have a `stop()` function that closes sockets, and a `join()` function that waits for `io_context.run()` to complete on all of its threads - that drains the `io_context` work queue. if `stop()` happens in the middle of a call to `process_request()`, that call will keep running until it hits a socket error, then follow its normal error path until the coroutine completes

in contrast, you might instead call `io_context.stop()` to cause all of the `io_context.run()` threads to return immediately without draining. here we still need to rely on RAII to guarantee that all of the memory gets released, but we won't have a chance to do any async cleanup. this was the strategy you took in https://github.com/ceph/ceph/blob/main/src/rgw/rgw_notify.cc#L467-L470 and i think it works fine there

**Comment by yuvalif** [comment]
beast + cpp20 coroutines examples were recently added. see: https://github.com/boostorg/beast/pull/2538

**Comment by yuvalif** [comment]
closing in favor of PR #49845

**Pull request closed without merging**